### PR TITLE
[CARBONDATA-2309][DataLoad] Add strategy to generate bigger carbondata files in case of small amo…

### PR DESCRIPTION
**Pull request by ndwangsen**
In some scenario, the input amount of loading data is small, but carbondata still distribute them to each executors (nodes) to do local-sort, thus resulting to small carbondata files generated by each executor. 

In  some extreme conditions, if the cluster is big enough or if the amount of data is small enough, the carbondata file contains only one blocklet or page.

I  think a new strategy should be introduced to solve the above problem.

The new strategy should:

be able to control the minimum amount of input data for each node
ignore data locality otherwise it may always choose a small portion of particular nodes

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [x] Any interfaces changed?
   NO
 - [x] Any backward compatibility impacted?
   NO
 - [x] Document update required?
   YES
 - [x] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
           YES
        - How it is tested? Please attach test report.
          Tested in local
        - Is it a performance related change? Please attach the performance test report.mac
          After this PR, performance is as we expected
        - Any additional information to help reviewers in testing this change.
          NO
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
          NO

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/4843/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/6002/

**Comment by kumarvishal09** [comment]
@ndwangsen This pr will help during concurrent query performance as number of task will be less,  but it will impact data loading performance as it will not use all the nodes and data locality is not considered. 

But adding a property is not the correct way. It's better to expose one data load strategy interface and add multiple implementation..like distribute the data loading across nodes or your pr scenario when user wants to distribute the data based on size. Please expose one data loading strategy interface and  concrete implementation for each type (existing+ size based distribution) and user can configure or pass in load parameter which strategy they want to opt based on their use case. If user is not passing any strategy implementation, it should take default implementation.  By this way user can add some custom strategy based on their use case

**Comment by kumarvishal09** [comment]
@ndwangsen What about the existing strategy???
 Can u please explain how u are planing to handle this??

**Comment by ndwangsen** [comment]
@kumarvishal09  If the user specifies or default that the minimum data load of the node is less than the average data amount of each node, the existing strategy is used to handle

File: processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java [hunk 1]
```diff
@@ -596,20 +599,50 @@ public static Dictionary getDictionary(AbsoluteTableIdentifier absoluteTableIden
...
+      }
+      // If the average expected size for each node greater than load min size,
+      // then fall back to default strategy
+      if (iLoadMinSize * 1024 * 1024 < sizePerNode) {
+        if (CarbonProperties.getInstance().isLoadSkewedDataOptimizationEnabled()) {
+          blockAssignmentStrategy = BlockAssignmentStrategy.BLOCK_SIZE_FIRST;
+        } else {
+          blockAssignmentStrategy = BlockAssignmentStrategy.BLOCK_NUM_FIRST;
+        }
+      } else {
```
Comment (xuchuanyin) [inline processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java]: Better to add log
```
LOG.info("Specified minimum data size to load is less than the average size for each node, fallback to default strategy" + blockAssignmentStrategy);
```

File: processing/src/main/java/org/apache/carbondata/processing/loading/model/LoadOption.java [hunk 2]
```diff
@@ -196,6 +196,8 @@
     optionsFinal.put("single_pass", String.valueOf(singlePass));
     optionsFinal.put("sort_scope", "local_sort");
     optionsFinal.put("sort_column_bounds", Maps.getOrDefault(options, "sort_column_bounds", ""));
+    optionsFinal.put(CarbonCommonConstants.CARBON_LOAD_MIN_SIZE_INMB,Maps.getOrDefault(options,CarbonCommonConstants.CARBON_LOAD_MIN_SIZE_INMB, CarbonCommonConstants
```
Comment (xuchuanyin) [inline processing/src/main/java/org/apache/carbondata/processing/loading/model/LoadOption.java]: add a space after `,`

File: processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java [hunk 3]
```diff
@@ -575,11 +577,12 @@ public static Dictionary getDictionary(AbsoluteTableIdentifier absoluteTableIden
    * @param noOfNodesInput -1 if number of nodes has to be decided
    *                       based on block location information
    * @param blockAssignmentStrategy strategy used to assign blocks
+   * @param loadMinSize the property load_min_size_inmb specified by the user
    * @return a map that maps node to blocks
    */
   public static Map<String, List<Distributable>> nodeBlockMapping(
       List<Distributable> blockInfos, int noOfNodesInput, List<String> activeNodes,
-      BlockAssignmentStrategy blockAssignmentStrategy) {
+      BlockAssignmentStrategy blockAssignmentStrategy, String loadMinSize ) {
```
Comment (xuchuanyin) [inline processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java]: How about changing the name `loadMinSize` to `expectedMinSizePerNode`?

**Comment by xuchuanyin** [comment]
@ndwangsen please resolve the conflicts and review comments

File: processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java (hunk 3 above)
Comment (ndwangsen) [reply processing/src/main/java/org/apache/carbondata/processing/util/CarbonLoaderUtil.java]: ok，I modify it according to your reviewed message，thanks.

**Comment by kumarvishal09** [comment]
@ndwangsen I think still load is based on property....and based on enum you are checking which type of strategy it will choose , as already mentioned in earlier comment 
Please expose one data loading strategy interface and add concrete implementation for each type (existing+ size based distribution) and user can configure or pass in load parameter which strategy they want to opt based on their use case. If user is not passing any strategy implementation, it should take default implementation. By this way user can add some custom strategy based on their use case**

**Comment by jackylk** [comment]
LGTM

**Pull request closed without merging**